quaddtype/numpy_quaddtype/src/scalar.c

```diff
@@ -624,6 +625,120 @@ static PyGetSetDef QuadPrecision_getset[] = {
     {NULL}  /* Sentinel */
 };
 
+/*
+ * Hash function for QuadPrecision scalars.
+ * 
+ * This implements the same algorithm as CPython's _Py_HashDouble, adapted for
+ * 128-bit floating point. The algorithm computes a hash based
+ * on the reduction of the value modulo the prime P = 2**PYHASH_BITS - 1.
+ * https://github.com/python/cpython/blob/20b69aac0d19a5e5358362410d9710887762f0e7/Python/pyhash.c#L87
+ * 
+ * Key invariant: hash(x) == hash(y) whenever x and y are numerically equal,
+ * even if x and y have different types. This ensures that:
+ *   hash(QuadPrecision(1.0)) == hash(1.0) == hash(1)
+ * 
+ * The algorithm:
+ * 1. Handle special cases: inf returns PYHASH_INF, nan uses pointer hash
+ * 2. Extract mantissa m in [0.5, 1.0) and exponent e via frexp(v) = m * 2^e
+ * 3. Process mantissa 28 bits at a time, accumulating into hash value x
+ * 4. Adjust for exponent using bit rotation (since 2^PYHASH_BITS ≡ 1 mod P)
+ * 5. Apply sign and handle the special case of -1 -> -2
+ */
+
+#if SIZEOF_VOID_P >= 8
+#  define PYHASH_BITS 61
+#else
+#  define PYHASH_BITS 31
+#endif
+#define PYHASH_MODULUS (((Py_uhash_t)1 << PYHASH_BITS) - 1)
+#define PYHASH_INF 314159
+
+static Py_hash_t
+QuadPrecision_hash(QuadPrecisionObject *self)
+{
+    Sleef_quad value;
+    int sign = 1;
+    
+    if (self->backend == BACKEND_SLEEF) {
+        value = self->value.sleef_value;
+    }
+    else {
+        value = Sleef_cast_from_doubleq1((double)self->value.longdouble_value);
+    }
+    
+    // Check for NaN - use pointer hash (each NaN instance gets unique hash)
+    // This prevents hash table catastrophic pileups from NaN instances
+    if (Sleef_iunordq1(value, value)) {
+        return _Py_HashPointer((void *)self);
```

Comment:
[Py_HashPointer](https://docs.python.org/3/c-api/hash.html#c.Py_HashPointer) was added in newer versions of the C API. You should probably vendor [pythoncapi-compat](https://github.com/python/pythoncapi-compat) and use the definition exposed by that header. That way when, in the future, CPython removes this symbol, your code won't break.

In general you should avoid directly using private symbols in the C API - they can change or break without notice, even in minor version bumps.